In [632]:
import pandas as pd
import numpy as np

loc = 'titanic3.xls'
dfraw=pd.read_excel(loc)

In [633]:
dfraw

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
1          1         1                     Allison, Master. Hudson Trevor   
2          1         0                       Allison, Miss. Helen Loraine   
3          1         0               Allison, Mr. Hudson Joshua Creighton   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5          1         1                                Anderson, Mr. Harry   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
7          1         0                             Andrews, Mr. Thomas Jr   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9          1         0                            Artagaveytia, Mr. Ramon   
10         1         0                             Astor, Col. John Jacob   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
13         1         1                       Barber, Miss. Ellen "Nellie"   
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
15         1         0                                Baumann, Mr. John D   
16         1         0                           Baxter, Mr. Quigg Edmond   
17         1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18         1         1                              Bazzani, Miss. Albina   
19         1         0                               Beattie, Mr. Thomson   
20         1         1                      Beckwith, Mr. Richard Leonard   
21         1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22         1         1                              Behr, Mr. Karl Howell   
23         1         1                              Bidois, Miss. Rosalie   
24         1         1                                  Bird, Miss. Ellen   
25         1         0                                Birnbaum, Mr. Jakob   
26         1         1                            Bishop, Mr. Dickinson H   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28         1         1                             Bissette, Miss. Amelia   
29         1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
...      ...       ...                                                ...   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1280       3         0                                    Vovk, Mr. Janko   
1281       3         0                               Waelens, Mr. Achille   
1282       3         0                                Ware, Mr. Frederick   
1283       3         0                        Warren, Mr. Charles William   
1284       3         0                                  Webber, Mr. James   
1285       3         0                                Wenzel, Mr. Linhart   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
1288       3         0                          Wiklund, Mr. Jakob Alfred   
1289       3         0                            Wiklund, Mr. Karl Johan   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1291       3         0                   Willer, Mr. Aaron ("Abi Weller")   
1292       3         0                                 Willey, Mr. Edward   
1293       3         0                  Williams, Mr. Howard Hugh "Harry"   
1294       3         0                               Williams, Mr. Leslie   
1295       3         0                                Windelov, Mr. Einar   
1296       3         0                                   Wirz, Mr. Albert   
1297       3         0                             Wiseman, Mr. Phillippe   
1298       3         0                          Wittevrongel, Mr. Camille   
12

# Data Cleaning

### This dataset needs dummies for pclass

In [634]:
dfraw['pclassone'] = np.where((dfraw['pclass']==1),1,0)
dfraw['pclasstwo'] = np.where((dfraw['pclass']==2),1,0)

### There are no missing values for pclass

In [635]:
len(dfraw[dfraw["pclass"].isnull()])

0

### This dataset needs some dummies for sex

In [636]:
dfraw['sexnum'] = np.where((dfraw["sex"]=="male"),1,0)

### Some research about the titanic tells me that cabins correspond to different levels of the ship, and the higher your class was the higher your cabin. To avoid collinearity, and to make my life easier, I am making a decision to leave cabin out as a predictor

In [637]:
len(dfraw[dfraw["home.dest"].isnull()])

564

### Similar things could be said about home destination, and since there are 564 missing destinations, this seems like a good variable to leave alone

In [638]:
len(dfraw[dfraw["embarked"].isnull()])

2

### Embarked has only two missing ports of departure, so dummies will be codes for the ports we have

In [639]:
pd.pivot_table(dfraw, index = "embarked", values = "survived")

survived
embarked          
C         0.555556
Q         0.357724
S         0.332604

### Here we can see there are three categories, so we need three dummies with the default being the two people who didn't have a port of departure

In [640]:
dfraw['portc'] = np.where((dfraw['embarked']=="C"),1,0)
dfraw['portq'] = np.where((dfraw['embarked']=="Q"),1,0)
dfraw['ports'] = np.where((dfraw['embarked']=="S"),1,0)

In [641]:
dfraw.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  pclassone  pclasstwo  sexnum  portc  \
0                     St Louis, MO          1          0       0      0   
1  Montreal, PQ / Chesterville, ON          1          0       1      0   
2  Montreal, PQ / Chesterville, ON          1          0       0      0   
3  Montreal, PQ / Chesterville, ON          1          0       1      0   
4  Montreal, PQ / Chesterville, ON          1          0       0      0   

   portq  ports  
0      0      1  
1      0      1  
2      0      1  
3      0      1  
4      0      1

In [642]:
# our df has so many columns its hard to see what we're doing
dfraw[['embarked', "ports", "portc", "portq"]].head(20)

embarked  ports  portc  portq
0         S      1      0      0
1         S      1      0      0
2         S      1      0      0
3         S      1      0      0
4         S      1      0      0
5         S      1      0      0
6         S      1      0      0
7         S      1      0      0
8         S      1      0      0
9         C      0      1      0
10        C      0      1      0
11        C      0      1      0
12        C      0      1      0
13        S      1      0      0
14        S      1      0      0
15        S      1      0      0
16        C      0      1      0
17        C      0      1      0
18        C      0      1      0
19        C      0      1      0

### Unsurpisingly 'body', which I believe is the person's weight, has so many missing values that it doesn't correlate with anything.

In [643]:
# There are 1188 missing observations under body
len(dfraw[dfraw["body"].isnull()])

1188

In [644]:
# out of a 1309 observation dataset
len(dfraw)

1309

### Is the survival probability for those with a missing bodyweight different than those with a recorded bodyweight?

In [645]:
dfraw["hasbody"] = np.where((dfraw["body"].isnull()),0,1)

pd.pivot_table(dfraw, index = "survived", values="hasbody")

hasbody
survived          
0         0.149567
1         0.000000

### None of the survivors had their body weights recorded, so it seems unlikely that bodyweight observations could be usefull in predicting survival.

In [646]:
len(dfraw[dfraw["age"].isnull()])

263

### age has 263 missing observations

In [647]:
dfraw["hasage"] = np.where(dfraw["age"].isnull(),0,1)
dfraw.loc[dfraw["age"].isnull()]

pd.pivot_table(dfraw, index = 'survived', values = "hasage")

hasage
survived          
0         0.765142
1         0.854000

### 76.5% of those who didn't survive had an age and 85.4% of those who did survive didn't have an age. Given that well over half of those who did and didn't make it both had a recorded age, it seems unlikely that removing age or replacing age with a mean age would be too problematic for our analysis

In [648]:
dfraw["meanage"] = dfraw.age.fillna(dfraw.age.mean())

### I've decided to add a new variable that replaces NaN observations for age with the mean age

### Fare has one nan, so lets drop it

In [649]:
#len(dfraw[dfraw["fare"].isnull()])

In [650]:
#dfraw = dfraw.dropna(subset = ["fare"], )

In [651]:
#len(dfraw[dfraw["fare"].isnull()])

In [652]:
len(dfraw[dfraw["sibsp"].isnull()])

0

In [653]:
len(dfraw[dfraw["parch"].isnull()])

0

### sibsp and parch have no NaN observations

In [654]:
len(dfraw[dfraw["sibsp"].isnull()])

0

In [655]:
len(dfraw[dfraw["parch"].isnull()])

0

### sibsp and parch have no NaN observations

### I'd also like to make a dummy variable for "hasfam" short for "has family" because having family on board might give you more will to live.

In [656]:
dfraw['hasfam'] = np.where(((dfraw["sibsp"] > 0) & (dfraw["parch"] > 0)),1,0)

### After the cleaning we have these new columns

In [657]:
dfraw

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
1          1         1                     Allison, Master. Hudson Trevor   
2          1         0                       Allison, Miss. Helen Loraine   
3          1         0               Allison, Mr. Hudson Joshua Creighton   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5          1         1                                Anderson, Mr. Harry   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
7          1         0                             Andrews, Mr. Thomas Jr   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9          1         0                            Artagaveytia, Mr. Ramon   
10         1         0                             Astor, Col. John Jacob   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
13         1         1                       Barber, Miss. Ellen "Nellie"   
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
15         1         0                                Baumann, Mr. John D   
16         1         0                           Baxter, Mr. Quigg Edmond   
17         1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18         1         1                              Bazzani, Miss. Albina   
19         1         0                               Beattie, Mr. Thomson   
20         1         1                      Beckwith, Mr. Richard Leonard   
21         1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22         1         1                              Behr, Mr. Karl Howell   
23         1         1                              Bidois, Miss. Rosalie   
24         1         1                                  Bird, Miss. Ellen   
25         1         0                                Birnbaum, Mr. Jakob   
26         1         1                            Bishop, Mr. Dickinson H   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28         1         1                             Bissette, Miss. Amelia   
29         1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
...      ...       ...                                                ...   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1280       3         0                                    Vovk, Mr. Janko   
1281       3         0                               Waelens, Mr. Achille   
1282       3         0                                Ware, Mr. Frederick   
1283       3         0                        Warren, Mr. Charles William   
1284       3         0                                  Webber, Mr. James   
1285       3         0                                Wenzel, Mr. Linhart   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
1288       3         0                          Wiklund, Mr. Jakob Alfred   
1289       3         0                            Wiklund, Mr. Karl Johan   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1291       3         0                   Willer, Mr. Aaron ("Abi Weller")   
1292       3         0                                 Willey, Mr. Edward   
1293       3         0                  Williams, Mr. Howard Hugh "Harry"   
1294       3         0                               Williams, Mr. Leslie   
1295       3         0                                Windelov, Mr. Einar   
1296       3         0                                   Wirz, Mr. Albert   
1297       3         0                             Wiseman, Mr. Phillippe   
1298       3         0                          Wittevrongel, Mr. Camille   
12

### The one variable I have overlooked so far is ticket, but I'm not sure I can cleanly extract what information is contained there, so I am simply leaving it out for the time being.

# Preliminary Analysis

### The variables we are left with to correlate with survived are: pclassone, pclasstwo, pclassthree(implied), sexnum, meanage, sibsp, parch, fare, portc, portq, noport(implied), ports and hasfam

In [658]:
dfraw[["survived", "pclassone", "pclasstwo", "sexnum", "meanage", "sibsp", "parch", "fare", "portc", "portq", "ports", "hasfam"]].corr()

survived  pclassone  pclasstwo    sexnum   meanage     sibsp  \
survived   1.000000   0.279449   0.050790 -0.528693 -0.050199 -0.027825   
pclassone  0.279449   1.000000  -0.296526 -0.107371  0.362587 -0.034256   
pclasstwo  0.050790  -0.296526   1.000000 -0.028862 -0.014193 -0.052419   
sexnum    -0.528693  -0.107371  -0.028862  1.000000  0.057398 -0.109609   
meanage   -0.050199   0.362587  -0.014193  0.057398  1.000000 -0.190747   
sibsp     -0.027825  -0.034256  -0.052419 -0.109609 -0.190747  1.000000   
parch      0.082660  -0.013033  -0.010057 -0.213125 -0.130872  0.373587   
fare       0.244265   0.600031  -0.121384 -0.185523  0.171892  0.160238   
portc      0.182123   0.325722  -0.134675 -0.066564  0.076179 -0.048396   
portq     -0.016071  -0.166101  -0.121973 -0.088651 -0.012718 -0.048678   
ports     -0.154558  -0.187353   0.197973  0.119504 -0.064267  0.075198   
hasfam     0.066122  -0.038107   0.017504 -0.138735 -0.220648  0.616986   

              parch      fare     portc     portq     ports    hasfam  
survived   0.082660  0.244265  0.182123 -0.016071 -0.154558  0.066122  
pclassone -0.013033  0.600031  0.325722 -0.166101 -0.187353 -0.038107  
pclasstwo -0.010057 -0.121384 -0.134675 -0.121973  0.197973  0.017504  
sexnum    -0.213125 -0.185523 -0.066564 -0.088651  0.119504 -0.138735  
meanage   -0.130872  0.171892  0.076179 -0.012718 -0.064267 -0.220648  
sibsp      0.373587  0.160238 -0.048396 -0.048678  0.075198  0.616986  
parch      1.000000  0.221539 -0.008635 -0.100943  0.073258  0.636915  
fare       0.221539  1.000000  0.286269 -0.130059 -0.172683  0.184786  
portc     -0.008635  0.286269  1.000000 -0.164166 -0.775441 -0.023282  
portq     -0.100943 -0.130059 -0.164166  1.000000 -0.489874 -0.088844  
ports      0.073258 -0.172683 -0.775441 -0.489874  1.000000  0.078436  
hasfam     0.636915  0.184786 -0.023282 -0.088844  0.078436  1.000000

### The correlations look non-zero for the most part. I will leave all of these variables in when running the first regression

# Regression Analysis

### Run a lineary probability model using the above coefficients

In [659]:
import statsmodels.formula.api as sm
reg1 = sm.ols(formula="survived ~ pclassone + pclasstwo + sexnum + meanage + sibsp + parch + fare + portc + portq + ports + hasfam", data =  dfraw).fit()
reg1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.373
Model:                            OLS   Adj. R-squared:                  0.368
Method:                 Least Squares   F-statistic:                     70.23
Date:                Fri, 12 Oct 2018   Prob (F-statistic):          2.78e-123
Time:                        23:41:31   Log-Likelihood:                -606.25
No. Observations:                1308   AIC:                             1237.
Df Residuals:                    1296   BIC:                             1299.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.9430      0.277      3.406      0.001       0.400       1.486
pclassone      0.3125      0.036      8.670      0.000       0.242       0.383
pclasstwo      0.1471      0.029      5.157      0.000       0.091       0.203
sexnum        -0.4949      0.024    -20.991      0.000      -0.541      -0.449
meanage       -0.0054      0.001     -5.821      0.000      -0.007      -0.004
sibsp         -0.0575      0.013     -4.347      0.000      -0.083      -0.032
parch         -0.0224      0.017     -1.353      0.176      -0.055       0.010
fare           0.0002      0.000      0.676      0.499      -0.000       0.001
portc         -0.1008      0.275     -0.366      0.714      -0.641       0.439
portq         -0.1749      0.277     -0.631      0.528      -0.718       0.369
ports         -0.1958      0.275     -0.712      0.477      -0.735       0.344
hasfam         0.1017      0.045      2.241      0.025       0.013       0.191
==============================================================================
Omnibus:                       52.080   Durbin-Watson:                   1.813
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               57.746
Skew:                           0.515   Prob(JB):                     2.89e-13
Kurtosis:                       3.001   Cond. No.                     3.33e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.33e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### The coefficient of fare seems rather small, and perhaps is not suiting our analysis well. Instead lets make it a log fare and see what a percent increase in fare amount does to your chances of survival on the titanic

In [660]:
# np.log(df['norm'])
dfraw["logfare"] = np.log(dfraw["fare"])

C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in log
  


### Apparently someone paid a zero dollar fare, that doesn't help

In [661]:
dfraw.loc[dfraw["fare"]==0]

pclass  survived                                   name   sex   age  \
7          1         0                 Andrews, Mr. Thomas Jr  male  39.0   
70         1         0  Chisholm, Mr. Roderick Robert Crispin  male   NaN   
125        1         0                       Fry, Mr. Richard  male   NaN   
150        1         0                  Harrison, Mr. William  male  40.0   
170        1         1                Ismay, Mr. Joseph Bruce  male  49.0   
223        1         0          Parr, Mr. William Henry Marsh  male   NaN   
234        1         0        Reuchlin, Jonkheer. John George  male  38.0   
363        2         0                  Campbell, Mr. William  male   NaN   
384        2         0         Cunningham, Mr. Alfred Fleming  male   NaN   
410        2         0       Frost, Mr. Anthony Wood "Archie"  male   NaN   
473        2         0                   Knight, Mr. Robert J  male   NaN   
528        2         0            Parkes, Mr. Francis "Frank"  male   NaN   
581        2         0             Watson, Mr. Ennis Hastings  male   NaN   
896        3         0                    Johnson, Mr. Alfred  male  49.0   
898        3         0        Johnson, Mr. William Cahoone Jr  male  19.0   
963        3         0                    Leonard, Mr. Lionel  male  36.0   
1254       3         1           Tornquist, Mr. William Henry  male  25.0   

      sibsp  parch  ticket  fare        cabin   ...    pclasstwo sexnum  \
7         0      0  112050   0.0          A36   ...            0      1   
70        0      0  112051   0.0          NaN   ...            0      1   
125       0      0  112058   0.0         B102   ...            0      1   
150       0      0  112059   0.0          B94   ...            0      1   
170       0      0  112058   0.0  B52 B54 B56   ...            0      1   
223       0      0  112052   0.0          NaN   ...            0      1   
234       0      0   19972   0.0          NaN   ...            0      1   
363       0      0  239853   0.0          NaN   ...            1      1   
384       0      0  239853   0.0          NaN   ...            1      1   
410       0      0  239854   0.0          NaN   ...            1      1   
473       0      0  239855   0.0          NaN   ...            1      1   
528       0      0  239853   0.0          NaN   ...            1      1   
581       0      0  239856   0.0          NaN   ...            1      1   
896       0      0    LINE   0.0          NaN   ...            0      1   
898       0      0    LINE   0.0          NaN   ...            0      1   
963       0      0    LINE   0.0          NaN   ...            0      1   
1254      0      0    LINE   0.0          NaN   ...            0      1   

      portc portq  ports  hasbody  hasage    meanage  hasfam  logfare  
7         0     0      1        0       1  39.000000       0     -inf  
70        0     0      1        0       0  29.881135       0     -inf  
125       0     0      1        0       0  29.881135       0     -inf  
150       0     0      1        1       1  40.000000       0     -inf  
170       0     0      1        0       1  49.000000       0     -inf  
223       0     0      1        0       0  29.881135       0     -inf  
234       0     0      1        0       1  38.000000       0     -inf  
363       0     0      1        0       0  29.881135       0     -inf  
384       0     0      1        0       0  29.881135       0     -inf  
410       0     0      1        0       0  29.881135       0     -inf  
473       0     0      1        0       0  29.881135       0     -inf  
528       0     0      1        0       0  29.881135       0     -inf  
581       0     0      1        0       0  29.881135       0     -inf  
896       0     0      1        0       1  49.000000       0     -inf  
898       0     0      1        0       1  19.000000       0     -inf  
963       0     0      1        0       1  36.000000       0     -inf  
1254      0     0      1        0       1  2

In [662]:
dfraw = dfraw.replace([-np.inf,], 0)

In [663]:
dfraw.fare.loc[dfraw["logfare"]==0] 

7       0.0
70      0.0
125     0.0
150     0.0
170     0.0
223     0.0
234     0.0
363     0.0
384     0.0
410     0.0
473     0.0
528     0.0
581     0.0
896     0.0
898     0.0
963     0.0
1254    0.0
Name: fare, dtype: float64

In [664]:
dfraw

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
1          1         1                     Allison, Master. Hudson Trevor   
2          1         0                       Allison, Miss. Helen Loraine   
3          1         0               Allison, Mr. Hudson Joshua Creighton   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5          1         1                                Anderson, Mr. Harry   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
7          1         0                             Andrews, Mr. Thomas Jr   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9          1         0                            Artagaveytia, Mr. Ramon   
10         1         0                             Astor, Col. John Jacob   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
13         1         1                       Barber, Miss. Ellen "Nellie"   
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
15         1         0                                Baumann, Mr. John D   
16         1         0                           Baxter, Mr. Quigg Edmond   
17         1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18         1         1                              Bazzani, Miss. Albina   
19         1         0                               Beattie, Mr. Thomson   
20         1         1                      Beckwith, Mr. Richard Leonard   
21         1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22         1         1                              Behr, Mr. Karl Howell   
23         1         1                              Bidois, Miss. Rosalie   
24         1         1                                  Bird, Miss. Ellen   
25         1         0                                Birnbaum, Mr. Jakob   
26         1         1                            Bishop, Mr. Dickinson H   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28         1         1                             Bissette, Miss. Amelia   
29         1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
...      ...       ...                                                ...   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1280       3         0                                    Vovk, Mr. Janko   
1281       3         0                               Waelens, Mr. Achille   
1282       3         0                                Ware, Mr. Frederick   
1283       3         0                        Warren, Mr. Charles William   
1284       3         0                                  Webber, Mr. James   
1285       3         0                                Wenzel, Mr. Linhart   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
1288       3         0                          Wiklund, Mr. Jakob Alfred   
1289       3         0                            Wiklund, Mr. Karl Johan   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1291       3         0                   Willer, Mr. Aaron ("Abi Weller")   
1292       3         0                                 Willey, Mr. Edward   
1293       3         0                  Williams, Mr. Howard Hugh "Harry"   
1294       3         0                               Williams, Mr. Leslie   
1295       3         0                                Windelov, Mr. Einar   
1296       3         0                                   Wirz, Mr. Albert   
1297       3         0                             Wiseman, Mr. Phillippe   
1298       3         0                          Wittevrongel, Mr. Camille   
12

### It appears we have successfully replaced all of the -infinities with zero, so that we now have a proper variable for logfare, lets regress a second time

In [665]:
reg2 = sm.ols(formula="survived ~ pclassone + pclasstwo + sexnum + meanage + sibsp + parch + logfare + portc + portq + ports + hasfam", data =  dfraw).fit()
reg2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.374
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     70.52
Date:                Fri, 12 Oct 2018   Prob (F-statistic):          1.03e-123
Time:                        23:41:32   Log-Likelihood:                -605.24
No. Observations:                1308   AIC:                             1234.
Df Residuals:                    1296   BIC:                             1297.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.8623      0.282      3.062      0.002       0.310       1.415
pclassone      0.2746      0.044      6.188      0.000       0.188       0.362
pclasstwo      0.1326      0.030      4.388      0.000       0.073       0.192
sexnum        -0.4908      0.024    -20.679      0.000      -0.537      -0.444
meanage       -0.0054      0.001     -5.798      0.000      -0.007      -0.004
sibsp         -0.0640      0.014     -4.581      0.000      -0.091      -0.037
parch         -0.0281      0.017     -1.654      0.098      -0.061       0.005
logfare        0.0302      0.019      1.573      0.116      -0.007       0.068
portc         -0.0901      0.275     -0.328      0.743      -0.630       0.449
portq         -0.1595      0.277     -0.576      0.565      -0.703       0.384
ports         -0.1815      0.275     -0.660      0.509      -0.721       0.358
hasfam         0.1013      0.045      2.235      0.026       0.012       0.190
==============================================================================
Omnibus:                       51.895   Durbin-Watson:                   1.815
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               57.530
Skew:                           0.514   Prob(JB):                     3.22e-13
Kurtosis:                       2.994   Cond. No.                     1.68e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.68e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### The p value is much improved, but it's still not significant. We can also try a squared fare

In [666]:
dfraw["faresquare"] = dfraw["fare"]*dfraw["fare"]

In [667]:
reg3 = sm.ols(formula="survived ~ pclassone + pclasstwo + sexnum + meanage + sibsp + parch + faresquare + portc + portq + ports + hasfam", data =  dfraw).fit()
reg3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.374
Model:                            OLS   Adj. R-squared:                  0.368
Method:                 Least Squares   F-statistic:                     70.29
Date:                Fri, 12 Oct 2018   Prob (F-statistic):          2.27e-123
Time:                        23:41:32   Log-Likelihood:                -606.05
No. Observations:                1308   AIC:                             1236.
Df Residuals:                    1296   BIC:                             1298.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.9490      0.277      3.427      0.001       0.406       1.492
pclassone      0.3180      0.031     10.139      0.000       0.257       0.380
pclasstwo      0.1485      0.028      5.228      0.000       0.093       0.204
sexnum        -0.4954      0.024    -21.066      0.000      -0.542      -0.449
meanage       -0.0054      0.001     -5.832      0.000      -0.007      -0.004
sibsp         -0.0566      0.013     -4.303      0.000      -0.082      -0.031
parch         -0.0218      0.016     -1.334      0.183      -0.054       0.010
faresquare  6.213e-07   6.68e-07      0.931      0.352   -6.88e-07    1.93e-06
portc         -0.1049      0.275     -0.381      0.703      -0.645       0.435
portq         -0.1787      0.277     -0.645      0.519      -0.722       0.365
ports         -0.1995      0.275     -0.725      0.468      -0.739       0.340
hasfam         0.1018      0.045      2.246      0.025       0.013       0.191
==============================================================================
Omnibus:                       52.133   Durbin-Watson:                   1.811
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               57.804
Skew:                           0.515   Prob(JB):                     2.81e-13
Kurtosis:                       3.005   Cond. No.                     9.12e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.12e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### This specification isn't yield results that are as good as I would have hoped, lets stick with logfare, and this time remove the port of embarkation

In [668]:
reg4 = sm.ols(formula="survived ~ pclassone + pclasstwo + sexnum + meanage + sibsp + parch + logfare + hasfam", data =  dfraw).fit()
reg4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.369
Model:                            OLS   Adj. R-squared:                  0.365
Method:                 Least Squares   F-statistic:                     95.11
Date:                Fri, 12 Oct 2018   Prob (F-statistic):          2.10e-124
Time:                        23:41:32   Log-Likelihood:                -610.51
No. Observations:                1308   AIC:                             1239.
Df Residuals:                    1299   BIC:                             1286.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.6851      0.055     12.411      0.000       0.577       0.793
pclassone      0.2838      0.044      6.456      0.000       0.198       0.370
pclasstwo      0.1206      0.030      4.073      0.000       0.063       0.179
sexnum        -0.4949      0.023    -21.104      0.000      -0.541      -0.449
meanage       -0.0055      0.001     -5.878      0.000      -0.007      -0.004
sibsp         -0.0684      0.014     -4.907      0.000      -0.096      -0.041
parch         -0.0316      0.017     -1.861      0.063      -0.065       0.002
logfare        0.0389      0.019      2.041      0.041       0.002       0.076
hasfam         0.1022      0.045      2.249      0.025       0.013       0.191
==============================================================================
Omnibus:                       50.264   Durbin-Watson:                   1.810
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               55.542
Skew:                           0.505   Prob(JB):                     8.69e-13
Kurtosis:                       2.982   Cond. No.                         204.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Most of our predictors are now calibrated to be significant within this linear probability model. Our adjusted r-squared is still a bit low, but we can continue to look for predictors within the data. One specification we can try is including an agesquare coefficient

In [669]:
dfraw["agesquare"] = dfraw["meanage"]*dfraw["meanage"]

In [670]:
reg5 = sm.ols(formula="survived ~ pclassone + pclasstwo + sexnum + meanage + agesquare + sibsp + parch + logfare + hasfam", data =  dfraw).fit()
reg5.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.372
Model:                            OLS   Adj. R-squared:                  0.368
Method:                 Least Squares   F-statistic:                     85.39
Date:                Fri, 12 Oct 2018   Prob (F-statistic):          1.61e-124
Time:                        23:41:33   Log-Likelihood:                -607.89
No. Observations:                1308   AIC:                             1236.
Df Residuals:                    1298   BIC:                             1288.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.7797      0.069     11.307      0.000       0.644       0.915
pclassone      0.2765      0.044      6.282      0.000       0.190       0.363
pclasstwo      0.1171      0.030      3.956      0.000       0.059       0.175
sexnum        -0.4957      0.023    -21.173      0.000      -0.542      -0.450
meanage       -0.0118      0.003     -4.038      0.000      -0.017      -0.006
agesquare   9.521e-05   4.17e-05      2.283      0.023    1.34e-05       0.000
sibsp         -0.0689      0.014     -4.951      0.000      -0.096      -0.042
parch         -0.0329      0.017     -1.942      0.052      -0.066       0.000
logfare        0.0389      0.019      2.045      0.041       0.002       0.076
hasfam         0.0814      0.046      1.758      0.079      -0.009       0.172
==============================================================================
Omnibus:                       48.704   Durbin-Watson:                   1.806
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.573
Skew:                           0.496   Prob(JB):                     2.33e-12
Kurtosis:                       3.017   Cond. No.                     9.68e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.68e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### This did nothing for us...

# Regression Results & Modeling

In [685]:
reg4.params

Intercept    0.685105
pclassone    0.283839
pclasstwo    0.120577
sexnum      -0.494875
meanage     -0.005455
sibsp       -0.068376
parch       -0.031563
logfare      0.038885
hasfam       0.102212
dtype: float64

In [686]:
betas = reg4.params

In [687]:
test = dfraw.loc[1226,["pclassone", "pclasstwo", "sexnum", "meanage", "sibsp", "parch", "logfare", "hasfam"]]

In [688]:
test

pclassone          0
pclasstwo          0
sexnum             1
meanage           19
sibsp              0
parch              0
logfare      2.06633
hasfam             0
Name: 1226, dtype: object

In [689]:
predict(test)

-0.5181752244822657

In [690]:
def predict(x):
    y=x*betas
    z=y.sum()
    return(z)

In [691]:
dfraw["chanceofsurvival"] = 0
for i, row in enumerate(dfraw.values):
                       dfraw.loc[i,["chanceofsurvival"]] = predict(dfraw.loc[i,["pclassone", "pclasstwo", "sexnum", "meanage", "sibsp", "parch", "logfare", "hasfam"]])

In [679]:
dfraw

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
1          1         1                     Allison, Master. Hudson Trevor   
2          1         0                       Allison, Miss. Helen Loraine   
3          1         0               Allison, Mr. Hudson Joshua Creighton   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5          1         1                                Anderson, Mr. Harry   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
7          1         0                             Andrews, Mr. Thomas Jr   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9          1         0                            Artagaveytia, Mr. Ramon   
10         1         0                             Astor, Col. John Jacob   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
13         1         1                       Barber, Miss. Ellen "Nellie"   
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
15         1         0                                Baumann, Mr. John D   
16         1         0                           Baxter, Mr. Quigg Edmond   
17         1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18         1         1                              Bazzani, Miss. Albina   
19         1         0                               Beattie, Mr. Thomson   
20         1         1                      Beckwith, Mr. Richard Leonard   
21         1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22         1         1                              Behr, Mr. Karl Howell   
23         1         1                              Bidois, Miss. Rosalie   
24         1         1                                  Bird, Miss. Ellen   
25         1         0                                Birnbaum, Mr. Jakob   
26         1         1                            Bishop, Mr. Dickinson H   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28         1         1                             Bissette, Miss. Amelia   
29         1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
...      ...       ...                                                ...   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1280       3         0                                    Vovk, Mr. Janko   
1281       3         0                               Waelens, Mr. Achille   
1282       3         0                                Ware, Mr. Frederick   
1283       3         0                        Warren, Mr. Charles William   
1284       3         0                                  Webber, Mr. James   
1285       3         0                                Wenzel, Mr. Linhart   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
1288       3         0                          Wiklund, Mr. Jakob Alfred   
1289       3         0                            Wiklund, Mr. Karl Johan   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1291       3         0                   Willer, Mr. Aaron ("Abi Weller")   
1292       3         0                                 Willey, Mr. Edward   
1293       3         0                  Williams, Mr. Howard Hugh "Harry"   
1294       3         0                               Williams, Mr. Leslie   
1295       3         0                                Windelov, Mr. Einar   
1296       3         0                                   Wirz, Mr. Albert   
1297       3         0                             Wiseman, Mr. Phillippe   
1298       3         0                          Wittevrongel, Mr. Camille   
12

In [680]:
dfraw.sort_values("chanceofsurvival", ascending = False)

pclass  survived                                               name  \
2          1         0                       Allison, Miss. Helen Loraine   
193        1         1                  Madill, Miss. Georgette Alexandra   
195        1         1                              Maioni, Miss. Roberta   
55         1         1                          Carter, Miss. Lucile Polk   
73         1         1                               Cleaver, Miss. Alice   
97         1         1  Douglas, Mrs. Frederick Charles (Mary Helene B...   
190        1         1                      Longley, Miss. Gretchen Fiske   
24         1         1                                  Bird, Miss. Ellen   
302        1         1                                   Ward, Miss. Anna   
0          1         1                      Allen, Miss. Elisabeth Walton   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
153        1         1                     Hays, Miss. Margaret Bechstein   
159        1         1                       Hippach, Miss. Jean Gertrude   
32         1         1                            Bonnell, Miss. Caroline   
12         1         1                      Aubart, Mme. Leontine Pauline   
255        1         1                               Sagesser, Mlle. Emma   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
137        1         1                       Graham, Miss. Margaret Edith   
13         1         1                       Barber, Miss. Ellen "Nellie"   
187        1         1                          Lines, Miss. Mary Conover   
319        1         1                          Wilson, Miss. Helen Alice   
229        1         1  Penasco y Castellana, Mrs. Victor de Satode (M...   
199        1         1   Mayne, Mlle. Berthe Antonine ("Mrs de Villiers")   
250        1         1                         Ryerson, Miss. Emily Borie   
66         1         1                        Chaudanson, Miss. Victorine   
108        1         1                            Fleming, Miss. Margaret   
182        1         1                                LeRoy, Miss. Bertha   
129        1         1                               Geiger, Miss. Amalie   
88         1         1                               Daniels, Miss. Sarah   
102        1         1              Earnshaw, Mrs. Boulton (Olive Potter)   
...      ...       ...                                                ...   
963        3         0                                Leonard, Mr. Lionel   
833        3         0                            Green, Mr. George Henry   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
790        3         0                                  Elias, Mr. Joseph   
979        3         0                        Lundahl, Mr. Johan Svensson   
825        3         0                     Goodwin, Master. Harold Victor   
629        3         0                        Andersson, Mr. Anders Johan   
644        3         0         Asplund, Mr. Carl Oscar Vilhelm Gustafsson   
838        3         0                       Gustafsson, Mr. Johan Birger   
827        3         0                 Goodwin, Master. William Frederick   
1015       3         0                                   Meo, Mr. Alfonzo   
831        3         0                     Goodwin, Mr. Charles Frederick   
830        3         0                        Goodwin, Mr. Charles Edward   
1152       3         0                            Robins, Mr. Alexander A   
736        3         0                                  Coxon, Mr. Daniel   
1179       3         0                              Sage, Mr. John George   
848        3         0                            Hansen, Mr. Claus Peter   
837        3         0                     Gustafsson, Mr. Anders Vilhelm   
1068       3         0                          Nysveen, Mr. Johan Hansen   
1274       3         0                          Vander Planke, Mr. Julius   
89

In [681]:
dfraw["somechance"] = np.where((dfraw['chanceofsurvival']>0),1,0)

In [682]:
dfraw[["survived", "somechance"]].head()

survived  somechance
0         1           1
1         1           0
2         0           1
3         0           0
4         0           1

# Results

In [683]:
pd.pivot_table(dfraw, values="survived", index="somechance")

survived
somechance          
0           0.257089
1           0.908367

### 90.8% of the people who my model predicted had some non-zero chance of survival, actually survived

In [684]:
pd.pivot_table(dfraw, values="somechance", index="survived")

somechance
survived            
0            0.02843
1            0.45600

### Of the people who did not survive, my model incorrectly predicted 2.8% of those people surviving

### of the people who did survive, my model only predicted 45.6% of their survival. So my model was only able to correctly predict 45% of the survivors